[Explicit Module Builds] Fallback on implicit module builds by filtering out Explicit Module clang arguments from invocation

artemcm opened the pull request:
When loading compilation flags from a serialized AST, filter out explicit module build commands for the time being and fallback on implicit module loading. In the future, we will utilize explicit dependency info, but for the time being the module loaders are not configured to take advantage of them.

Resolves rdar://121078994

artemcm committed:
[Explicit Module Builds] Fallback on implicit module builds by filtering out Explicit Module clang arguments from invocation

When loading compilation flags from a serialized AST, filter out explicit module build commands for the time being and fallback on implicit module loading.
In the future, we will utilize explicit dependency info, but for the time being the module loaders are not configured to take advantage of them.

Resolves rdar://121078994

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1294,6 +1302,29 @@ static bool DeserializeAllCompilerFlags(swift::CompilerInvocation &invocation,
+        if (llvm::find(invocation.getExtraClangArgs(), "-fno-implicit-modules")) {
+          auto matchExplicitBuildOption = [](StringRef &arg) {
+            for (const auto &option : knownExplicitModulePrefixes)
+              if (arg.consume_front(option))
+                return true;
```
It seems this doesn't need to be mutating the arguments, but maybe I've missed something?

```suggestion
              if (arg.starts_with(option))
                return true;
```

artemcm replied:
Changed to `starts_with`, thanks!

kastiglione reviewed (commented):
Thanks! I have one implementation detail question/comment.

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test

kastiglione commented:
I can repro the test failure locally. I'm investigating.

kastiglione commented:
https://github.com/apple/llvm-project/pull/8017

@swift-ci test

kastiglione commented:
https://github.com/apple/llvm-project/pull/8025

@swift-ci test

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1294,6 +1302,29 @@ static bool DeserializeAllCompilerFlags(swift::CompilerInvocation &invocation,
+        // module loading.
+        // TODO: Incorporate loading explicit module dependencies to
+        // speedup dependency resolution.
+        if (llvm::find(invocation.getExtraClangArgs(), "-fno-implicit-modules")) {
```
Out of curiosity, would `SwiftASTContext::AddExtraClangArgs` be a more canonical place to filter this out, since we already do filtering there?(compare to https://github.com/apple/llvm-project/pull/8025)

adrian-prantl replied:
Are you still planning to do this, or is it not a good idea for reasons? (Doing it in a follow-up is fine)

artemcm replied:
I am going to take a look at moving this to `AddExtraClangArgs` now, thanks.

artemcm replied:
Updated the commit to filter in `SwiftASTContext::AddExtraClangArgs`, thanks @adrian-prantl .

kastiglione committed:
[lldb] Ignore -triple flag for ClangImporter (#8025)

This change is in support of explicit swift modules (apple/llvm-project#8017).

With explicit modules, the resulting clang flags include the `-triple` flag, which has
not been supported by `SwiftASTContext::AddExtraClangArgs`. This change ignores any
found triple, relying on lldb's existing triple/target discovery functionality.

kastiglione commented:
@swift-ci test

kastiglione commented:
@swift-ci test windows

kastiglione commented:
@swift-ci test macOS

kastiglione commented:
🎉

artemcm commented:
@swift-ci test

felipepiovezan commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1159,6 +1159,14 @@ static std::string GetPluginServerForSDK(llvm::StringRef sdk_path) {
+        "-fmodule-file=",
+        "-fno-implicit-modules",
+        "-fno-implicit-module-maps"};
+}
```
FWIW `static` and anonymous namespaces are the same thing: both have internal linkage and static storage duration

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1159,6 +1159,14 @@ static std::string GetPluginServerForSDK(llvm::StringRef sdk_path) {
 }
 
+namespace {
+  static constexpr std::array<std::string_view, 4> knownExplicitModulePrefixes =
```
```suggestion
  static constexpr std::array<std::string_view, 4> g_known_eplicit_module_prefixes =
```
to make this conform to LLDB naming conventions.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1518,7 +1527,23 @@ void SwiftASTContext::AddExtraClangArgs(const std::vector<std::string> &source,
   llvm::SmallString<128> cur_working_dir;
   llvm::SmallString<128> clang_argument;
+
+  auto matchExplicitBuildOption = [](StringRef arg) {
```
```suggestion
  auto match_explicit_build_option = [](StringRef arg) {
```

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test macOS platform

artemcm commented:
@swift-ci test Windows platform

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test Windows platform

artemcm commented:
@swift-ci test macOS platform

kastiglione commented:
@swift-ci test windows
